nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -134,6 +134,10 @@ class BtActionNode : public BT::ActionNodeBase
       on_tick();
 
       on_new_goal_received();
+
+      rclcpp::spin_some(node_);
+
+      return BT::NodeStatus::RUNNING;
     }
```

Comment:
@SteveMacenski 
`BtActionNode` has `spin_some` already in the next `if` clause, which is called in the previous version.
I added a return statement here, so I think another `spin_some` is required.

Reply:
Ah, I think that return should be removed as well. Mostly because that really changes the default behavior of things. This case seems reasonable, but then the first time we tick a node, the `on_wait_for_result` function won't be  called, which is pretty instrumental for processing updates to the follow path task and another other BT nodes that have preemptable action servers. While its atypical for there to be an update so quickly, there could be one from external stimulii that would be good to check for on the first iteration. 

I see your point though, this does seem more clear and correct. What do you think?

Reply:
This change is for the test. Many tests assume that initial tick() returns `RUNNING`.
Without these lines, it could return `SUCCESS` in the very first tick(). 

```
./test/plugins/action/test_navigate_to_pose_action.cpp:127:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_navigate_to_pose_action.cpp:145:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_follow_path_action.cpp:130:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_follow_path_action.cpp:150:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_compute_path_to_pose_action.cpp:132:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_compute_path_to_pose_action.cpp:156:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_spin_action.cpp:142:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_back_up_action.cpp:144:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
./test/plugins/action/test_wait_action.cpp:143:  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::RUNNING);
```

Reply:
> This case seems reasonable, but then the first time we tick a node, the `on_wait_for_result` function won't be called, which is pretty instrumental for processing updates to the follow path task and another other BT nodes that have preemptable action servers.

- `on_wait_for_result` is only implemented by `FollowPathAction`
- the `on_wait_for_result` will do nothing in the first tick() as `goal_.path` is identical to `new_path`, because other nodes won't change `path` on the blackboard while this tick() is running
https://github.com/ros-planning/navigation2/blob/07c6b59098f9db2d064699fbee229a2dd2bb506d/nav2_behavior_tree/plugins/action/follow_path_action.cpp#L39-L48

Reply:
I'm fine with this, makes complete sense to me.

Reply:
https://github.com/ros-planning/navigation2/blob/bb29d801d504cf5de8c33524d01f76974f76e23a/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L161

Its still going to return running, its been this way for months, I'm confused as to what that is fixing? Either way you spin and return running. Its valid for a node to return SUCCESS or FAILURE on a tick, I'm not sure why it wouldn't be. Decorator and other action nodes like clearing costmaps do that.

Reply:
Everything else in this PR looks good

Reply:
I believe tree spins with 10ms timeout, meaning 100Hz.
https://github.com/ros-planning/navigation2/blob/2e0697c64c265f31bd4ef6c5ed43374841242077/nav2_behavior_tree/include/nav2_behavior_tree/behavior_tree_engine.hpp#L41
https://github.com/ros-planning/navigation2/blob/2e0697c64c265f31bd4ef6c5ed43374841242077/nav2_behavior_tree/src/behavior_tree_engine.cpp#L44

Also, I think 10ms is short enough for actions by its description
> Actions are useful when a response may take a significant length of time. They allow a client to track the progress of a request, get the final outcome, and optionally cancel the request before it completes.

http://design.ros2.org/articles/actions.html

Reply:
So it seems, thanks for the correction. Doesn't change the core of the argument; this adds latency to fast running actions

Reply:
> this adds latency to fast running actions
I understand the point, but then I came up with a question, why is `ComputePathToPose` action?
Other actions involve "robot" action like NavigateToPose, Spin, and Wait and I don't think it will finish in 10ms. `ComputePathToPose` does not have feedback and it looks like a service.

Anyway, this is off-topic from this PR, I will revert the last commit.

Reply:
Thanks, yeah I agree its probably not an issue, but since this library is external facing, I have to be especially careful about considering atypical uses